File: code/nest/lab4/l4e3_condition_kernel.cc
#include <cstring>
#include "system.h"
#include "synch.h"

#ifndef THREAD_NUM
#define THREAD_NUM 3
#endif

void work(int ticks) {
    ASSERT(interrupt->getLevel() == IntOn);
    for (int i = 0; i < ticks/10; i++) {
        interrupt->SetLevel(IntOff);
        interrupt->SetLevel(IntOn);
    }
}

Lock lock("l4e3_lock");
Condition condition("l4e3_condition");

void run(void *p)
{
    DEBUG('e', "%s begins\n", currentThread->getName());
    lock.Acquire();
    DEBUG('e', "%s acquired lock\n", currentThread->getName());
    DEBUG('e', "%s starts waiting\n", currentThread->getName());
    condition.Wait(&lock);
    DEBUG('e', "%s finished waiting\n", currentThread->getName());
    lock.Release();
    DEBUG('e', "%s released lock\n", currentThread->getName());
}

int Nest(void *arg) {
    DEBUG('e', "Entering Nest()\n");
    char* threadNames[THREAD_NUM+1];
    threadNames[0] = "main";

    for (int i = 1; i <= THREAD_NUM; i++) 
    {
        threadNames[i] = new char[10];
        sprintf(threadNames[i], "thread%d", i);
        Thread *t = new Thread(threadNames[i]);
        t->Fork(run, (void*)i);
    }

    work(500);
    lock.Acquire();
    DEBUG('e', "%s signaling one thread\n", currentThread->getName());
    condition.Signal(&lock);
    DEBUG('e', "%s signaled one thread\n", currentThread->getName());
    lock.Release();
    work(500);
    lock.Acquire();
    DEBUG('e', "%s broadcasting all threads\n", currentThread->getName());
    condition.Broadcast(&lock);
    DEBUG('e', "%s broadcasted all threads\n", currentThread->getName());
    lock.Release();

    return 0;
}

File: code/nest/lab4/l4e3_condition_kernel.py
import re
from collections import namedtuple
from collections import OrderedDict
from operator import attrgetter
# import sys
# raise Exception(str(sys.path))
from parse_result import *

def is_finished_waiting(line):
    a = re.findall("(.*) finished waiting\n", line)
    if (len(a) > 0):
        return a[0]
    else:
        return None

def is_signaling(line):
    a = re.findall("(.*) signaling one thread\n", line)
    if (len(a) > 0):
        return a[0]
    else:
        return None

def is_broadcasting(line):
    a = re.findall("(.*) broadcasting all threads\n", line)
    if (len(a) > 0):
        return a[0]
    else:
        return None

def check(lines):
    signaled = False
    signal = 0
    broadcast = False
    finished = 0
    THREAD_NUM = 3
    for line_num, line in enumerate(lines):
        ret = is_finished_waiting(line)
        if ret is not None:
            name = ret
            if not broadcast and signaled == False:
                error_message = 'Error in line {}: \
                    {} finished waiting, but condition variable hasn\'t signaled or broadcasted'\
                    .format(line_num+1, name)
                return False, error_message
            elif not broadcast and signal <= 0:
                error_message = 'Error in line {}: \
                    {} finished waiting, but condition variable hasn\'t broadcasted, and one thread has been signaled'\
                    .format(line_num+1, name)
                return False, error_message
            else:
                finished += 1
                signal -= 1
            continue

        ret = is_signaling(line)
        if ret is not None:
            signal += 1
            signaled = True
            continue
            
        ret = is_broadcasting(line)
        if ret is not None:
            broadcast = True
            continue
    
    if finished == THREAD_NUM:
        return True, None
    else:
        error_message = 'There are {} working threads, but {} finished working'\
            .format(THREAD_NUM, finished)
        return False, error_message
